File: sinopia/Response.js
import CBOR from '../other_libs/cbor';

class CborStructItem {

    constructor(...args) {
       this.setValue(...args);
    }

    setValue(...args) {
       this.record = args;
    }

    fromCBOR (ser_data) {
       var json_data = CBOR.decode(ser_data);
       this.fromData(json_data);
    }

    toCBOR () {
       var data = this.toData();
       return CBOR.encode(data);
    }

    toData () {
       return this.record;
    }

    fromData (json_data) {
       this.setValue(json_data);
    }
}

class CborEnumItem extends CborStructItem {
    setValue(...args) {
       if (arguments.length > 0) {
         this.selector = arguments[0];
         this.record = arguments.slice(0,1);
       }
    }

    toData () {
       return [this.selector, ...this.record];
    }
}

// response to file operations
class FileResponse extends CborEnumItem {
}

FileResponse.CurrentDBChanged = 0;   // the DB in use just changed
FileResponse.AvailableDBs = 1;   // list of DB's on disk
FileResponse.OpenedDB = 2;   // just opened the DB
FileResponse.CreatedDB = 3;   // just created the DB
FileResponse.DeletedDB = 4;   // just deleted the DB
FileResponse.SavedDB = 5;   // just saved the DB

// response to database queries
class QueryResponse extends CborEnumItem {
}

QueryResponse.DbInfo = 0;   // detailed DB info as JSON text
QueryResponse.DbQuery = 1;   // query answer as JSON text

// response to commands
class Response extends CborEnumItem {
    toData () {
       var arr_in = this.record.slice(); var arr_out = [];
       if (this.selector == 0) {
       }
       if (this.selector == 1) {
            arr_out.push(arr_in.shift().toData());
       }
       if (this.selector == 2) {
            arr_out.push(arr_in.shift().toData());
       }
       return [this.selector, ...arr_out];
    }

     fromData (json_data) {
       var arr_in = json_data.slice(0, 1); var arr_out = [];
       if (json_data[0] == 0) {
       }
       if (json_data[0] == 1) {
            arr_out.push((new FileResponse()).fromData(arr_in.shift());
       }
       if (json_data[0] == 2) {
            arr_out.push((new QueryResponse()).fromData(arr_in.shift());
       }
       this.selector = json_data[0];
       this.record = arr_out;
       }
}

Response.NoResponse = 0;   // response to NoOP command (is this needed?)
Response.ResFr = 1;   // response to a file command
Response.ResQuery = 2;   // response to a query command

File: sinopia/Response.hs
import Data.Binary.Serialise.CBOR
import Data.Binary.Serialise.CBOR.Encoding
import Data.Binary.Serialise.CBOR.Decoding

import Data.Text
import Data.Monoid
import Control.Applicative


-- | response to file operations
data FileResponse = CurrentDBChanged Text -- ^ the DB in use just changed 
    | AvailableDBs [Text] -- ^ list of DB's on disk 
    | OpenedDB Text -- ^ just opened the DB 
    | CreatedDB Text -- ^ just created the DB 
    | DeletedDB Text -- ^ just deleted the DB 
    | SavedDB Text -- ^ just saved the DB 
    deriving (Eq, Read, Show)

-- | response to database queries
data QueryResponse = DbInfo Text -- ^ detailed DB info as JSON text 
    | DbQuery Text -- ^ query answer as JSON text 
    deriving (Eq, Read, Show)

-- | response to commands
data Response = NoResponse -- ^ response to NoOP command (is this needed?) 
    | ResFr FileResponse -- ^ response to a file command 
    | ResQuery QueryResponse -- ^ response to a query command 
    | TestAnswer Text -- ^ arbitrary test answer as text 
    deriving (Eq, Read, Show)

instance Serialise FileResponse where
    encode (CurrentDBChanged v1) = encodeListLen 2 <>  encode (0::Int) <> encode v1
    encode (AvailableDBs v1) = encodeListLen 2 <>  encode (1::Int) <> encode v1
    encode (OpenedDB v1) = encodeListLen 2 <>  encode (2::Int) <> encode v1
    encode (CreatedDB v1) = encodeListLen 2 <>  encode (3::Int) <> encode v1
    encode (DeletedDB v1) = encodeListLen 2 <>  encode (4::Int) <> encode v1
    encode (SavedDB v1) = encodeListLen 2 <>  encode (5::Int) <> encode v1
    decode = do
        decodeListLen
        i <- decode :: Decoder s Int
        case i of
            0 -> (CurrentDBChanged <$> decode)
            1 -> (AvailableDBs <$> decode)
            2 -> (OpenedDB <$> decode)
            3 -> (CreatedDB <$> decode)
            4 -> (DeletedDB <$> decode)
            5 -> (SavedDB <$> decode)

instance Serialise QueryResponse where
    encode (DbInfo v1) = encodeListLen 2 <>  encode (0::Int) <> encode v1
    encode (DbQuery v1) = encodeListLen 2 <>  encode (1::Int) <> encode v1
    decode = do
        decodeListLen
        i <- decode :: Decoder s Int
        case i of
            0 -> (DbInfo <$> decode)
            1 -> (DbQuery <$> decode)

instance Serialise Response where
    encode (NoResponse) = encodeListLen 1 <>  encode (0::Int) 
    encode (ResFr v1) = encodeListLen 2 <>  encode (1::Int) <> encode v1
    encode (ResQuery v1) = encodeListLen 2 <>  encode (2::Int) <> encode v1
    encode (TestAnswer v1) = encodeListLen 2 <>  encode (3::Int) <> encode v1
    decode = do
        decodeListLen
        i <- decode :: Decoder s Int
        case i of
            0 -> (pure NoResponse)
            1 -> (ResFr <$> decode)
            2 -> (ResQuery <$> decode)
            3 -> (TestAnswer <$> decode)

